### PR TITLE
BUG: stats: Fix logpdf with large negative values for logistic and genlogistic dists.

### LINKED ISSUE #12490: BUG: stats: logistic and genlogistic logpdf overflow for large negative x.
The `_logpdf` method of the `logistic` distribution is implemented as 

    def _logpdf(self, x):
        return -x - 2. * sc.log1p(np.exp(-x))

This works fine for large positive x:

```
In [3]: from scipy.stats import logistic

In [4]: logistic.logpdf(800)
Out[4]: -800.0
```
But the term `exp(-x)` triggers an overflow with a large negative value:
```
In [5]: logistic.logpdf(-800)
/Users/warren/mc37scipy/lib/python3.7/site-packages/scipy-1.6.0.dev0+4c7b25e-py3.7-macosx-10.9-x86_64.egg/scipy/stats/_continuous_distns.py:4847: RuntimeWarning: overflow encountered in exp
  return -x - 2. * sc.log1p(np.exp(-x))
Out[5]: -inf
```
The distribution is symmetric, so the code should be able to handle both tails equally well.  Because of the symmetry, a fix should be as simple as setting `x = np.abs(x)` before computing the return values.

`genlogistic` has the same problem, but except when `c` is 1, it is not symmetric, so a little more care is needed to fix it.

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
Closes gh-12490.

### CODE REVIEW

File: scipy/stats/_continuous_distns.py [hunk 1]
```diff
@@ -2162,7 +2162,13 @@ def _pdf(self, x, c):
         return np.exp(self._logpdf(x, c))
 
     def _logpdf(self, x, c):
-        return np.log(c) - x - (c+1.0)*sc.log1p(np.exp(-x))
+        # Two mathematically equivalent expressions for log(pdf(x, c)):
+        #     log(pdf(x, c)) = log(c) - x - (c + 1)*log(1 + exp(-x))
+        #                    = log(c) + c*x - (c + 1)*log(1 + exp(x))
+        t = _lazywhere(x > 0, (x, c),
+                       f=lambda x, c: -x - (c + 1)*sc.log1p(np.exp(-x)),
+                       f2=lambda x, c: c*x - (c + 1)*sc.log1p(np.exp(x)))
```
Comment (sethtroisi) [inline scipy/stats/_continuous_distns.py:2170]: Free free to ignore:

I've always assumed `_lazywhere` is fairly slow. which seems to be generally true
```
In [18]: %timeit _lazywhere((t > 0), (t,), lambda v: sc.log1p(v), -np.inf)                                                            
46.3 µs ± 507 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [19]: %timeit sc.log1p(t)                                                                                                          
1.74 µs ± 35.1 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

At the cost of some ugliness it seems we can avoid it here
```
import numpy as np
import scipy.special as sc
from scipy._lib._util import _lazywhere

In [4]: c = 5 
In [5]: x = np.arange(-10, 10, 0.78) 
In [6]: def method1(x, c): 
   ...:    return _lazywhere(x > 0, (x, c), 
   ...:        f=lambda x, c: -x - (c + 1)*sc.log1p(np.exp(-x)), 
   ...:        f2=lambda x, c: c*x - (c + 1)*sc.log1p(np.exp(x))) 
   ...:                                                                                                                               
In [8]: method1(x, c)                                                                                                                 
Out[8]: 
array([-50.00027239, -46.1005942 , -42.20129616, -38.30282718,
       -34.4061657 , -30.51344214, -26.62928494, -22.76370109,
       -18.93810086, -15.19726975, -11.63049992,  -8.39895618,
        -5.74097906,  -3.89357109,  -2.93248309,  -2.70671618,
        -2.96252599,  -3.48601936,  -4.14466658,  -4.86820655,
        -5.62214626,  -6.39016213,  -7.16466052,  -7.94213686,
        -8.72097964,  -9.50044909])

In [10]: def method2(x, c): 
    ...:     mult = -(c-1) * (x < 0) - 1 
    ...:     t = np.abs(x) 
    ...:     return mult * t - (c+1) * sc.log1p(np.exp(-t)) 
    ...:                                                                                                                              

In [11]: method2(x, c)                                                                                                                
Out[11]: 
array([-50.00027239, -46.1005942 , -42.20129616, -38.30282718,
       -34.4061657 , -30.51344214, -26.62928494, -22.76370109,
       -18.93810086, -15.19726975, -11.63049992,  -8.39895618,
        -5.74097906,  -3.89357109,  -2.93248309,  -2.70671618,
        -2.96252599,  -3.48601936,  -4.14466658,  -4.86820655,
        -5.62214626,  -6.39016213,  -7.16466052,  -7.94213686,
        -8.72097964,  -9.50044909])

In [12]: %timeit method1(x, c)                                                                                                        
147 µs ± 4.59 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [13]: %timeit method2(x, c)                                                                                                        
16 µs ± 156 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

File: scipy/stats/tests/test_distributions.py [hunk 2]
```diff
@@ -855,6 +870,21 @@ def test_extreme_values(self):
         assert_allclose(stats.logistic.ppf(1 - p), desired)
         assert_allclose(stats.logistic.isf(p), desired)
 
+    def test_logpdf_basic(self):
+        logp = stats.logistic.logpdf([-15, 0, 10])
+        # Expected values were computed with mpmath, using 50 digits
```
Comment (sethtroisi) [inline scipy/stats/tests/test_distributions.py:875]: Nit: I'd love if the comment fit on one line like

`# Expected values were computed with mpmath @ 50 digits`
or
`        # Expected values were computed with mpmath using 50 digits of precision` 
which is maybe exactly 80

File: scipy/stats/tests/test_distributions.py [hunk 3]
```diff
@@ -675,6 +675,21 @@ def test_rvs_Generator(self):
                                 random_state=np.random.default_rng())
 
 
+class TestGenLogistic:
+
+    # Expected values were computed using mpmath with 50 digits
```
Comment (sethtroisi) [inline scipy/stats/tests/test_distributions.py:680]: Nit: no line wrap needed would be nice if it matched lower comment.

**Review by sethtroisi — APPROVED** [review]
LGTM otherwise

File: scipy/stats/_continuous_distns.py (hunk 1 above)
Comment (WarrenWeckesser) [reply scipy/stats/_continuous_distns.py:2170]: +1. It might be a bit uglier, but the speedup is worth it.

**Comment by WarrenWeckesser** [comment]
@sethtroisi thanks for the review.  I updated the `logpdf` code to avoid `_lazywhere`, and tweaked the comments while I was at it.

**Comment by sethtroisi** [comment]
@WarrenWeckesser looks good.
I'm trying to calibrate norms. Given this was open for several days with no other comment (and the changes after comments was \~small\~) would it be responsible to merge this immediately?

**Comment by WarrenWeckesser** [comment]
For a relatively small change such as this (especially when no API changes are involved), if you are satisfied that the change is correct, has appropriate unit tests, and conforms to PEP 8 and other SciPy coding guidelines, then yes, you can merge immediately.

**Comment by sethtroisi** [comment]
Thanks for the advice and the clean PR @WarrenWeckesser

**Comment by WarrenWeckesser** [comment]
Thanks @sethtroisi!

**Pull request merged**